doc/man/syslog-ng-query.1.in
'\" t
.\"     Title: syslog-ng-query
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 07/05/2016
.\"    Manual: The syslog-ng-query tool manual page
.\"    Source: 5 F6
.\"  Language: English
.\"
.TH "SYSLOG\-NG\-QUERY" "1" "07/05/2016" "5 F6" "The syslog-ng-query tool manua"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
syslog-ng-query \- Query metrics and statistic data from a running instance
.SH "SYNOPSIS"
.HP \w'\fBsyslog\-ng\-query\fR\ 'u
\fBsyslog\-ng\-query\fR [command] [options]
.SH "DESCRIPTION"
.PP
NOTE: The syslog\-ng\-query application is distributed with the system logging application, and is usually part of the syslog\-ng package\&. The latest version of the syslog\-ng application is available at
\m[blue]\fBthe official syslog\-ng website\fR\m[]\&.
.PP
This manual page is only an abstract; for the complete documentation of syslog\-ng, see
\m[blue]\fBThe syslog\-ng Administrator Guide\fR\m[]\&.
.PP
The
\fBsyslog\-ng\-query\fR
application is a utility that can be used to query metrics and statistic data from a running instance\&.
.PP
The application stores various data, metrics, and statistics in a hierarchical data structure tree (\fBsyslog\-ng\fR
is the root node of the tree)\&. The nodes of the tree can contain properties (but not every node contains properties)\&. Every property has a name and a value\&. For example:
.sp
.if n \{\
.RS 4
.\}
.nf
[syslog\-ng]
|
|_[destinations]\-[network]\-[tcp]\->[stats]\->{received:12;dropped:2}
|
|_[sources]\-[sql]\-[stats]\->{reveived:501;dropped:0}
.fi
.if n \{\
.RE
.\}
.PP
You can query the nodes of this tree, and also use filters to select the information you need\&. A query is actually a path in the tree\&. You can also use the
\fB?\fR
and
\fB*\fR
wildcards\&. For example:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Select every property:
\fB*\fR
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Select all
\fBdropped\fR
value from every
\fBstats\fR
node:
\fB*\&.stats\&.dropped\fR
.RE
.PP
The nodes and properties available in the tree depend on your configuration (that is, the sources, destinations, and other objects you have configured), and also on your
\fIstats\-level()\fR
settings\&.
.SH "THE LIST COMMAND"
.HP \w'\fBsyslog\-ng\-query\ list\fR\ 'u
\fBsyslog\-ng\-query list\fR
.PP
Use the
\fBsyslog\-ng\-query list\fR
command to display the list of metrics that collects about the processed messages\&. For details about the displayed metrics, see
\m[blue]\fBThe syslog\-ng Administrator Guide\fR\m[]???\&.
.PP
An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
stats
center\&.received\&.stats\&.processed
center\&.queued\&.stats\&.processed
destination\&.java\&.d_elastic#0\&.java_dst(ElasticSearch,elasticsearch\-syslog\-ng\-test,t7cde889529c034aea9ec_micek)\&.stats\&.dropped
destination\&.java\&.d_elastic#0\&.java_dst(ElasticSearch,elasticsearch\-syslog\-ng\-test,t7cde889529c034aea9ec_micek)\&.stats\&.processed
destination\&.java\&.d_elastic#0\&.java_dst(ElasticSearch,elasticsearch\-syslog\-ng\-test,t7cde889529c034aea9ec_micek)\&.stats\&.stored
destination\&.d_elastic\&.stats\&.processed
source\&.s_tcp\&.stats\&.processed
source\&.severity\&.7\&.stats\&.processed
source\&.severity\&.0\&.stats\&.processed
source\&.severity\&.1\&.stats\&.processed
source\&.severity\&.2\&.stats\&.processed
source\&.severity\&.3\&.stats\&.processed
source\&.severity\&.4\&.stats\&.processed
source\&.severity\&.5\&.stats\&.processed
source\&.severity\&.6\&.stats\&.processed
source\&.facility\&.7\&.stats\&.processed
source\&.facility\&.16\&.stats\&.processed
source\&.facility\&.8\&.stats\&.processed
source\&.facility\&.17\&.stats\&.processed
source\&.facility\&.9\&.stats\&.processed
source\&.facility\&.18\&.stats\&.processed
source\&.facility\&.19\&.stats\&.processed
source\&.facility\&.20\&.stats\&.processed
source\&.facility\&.0\&.stats\&.processed
source\&.facility\&.21\&.stats\&.processed
source\&.facility\&.1\&.stats\&.processed
source\&.facility\&.10\&.stats\&.processed
source\&.facility\&.22\&.stats\&.processed
source\&.facility\&.2\&.stats\&.processed
source\&.facility\&.11\&.stats\&.processed
source\&.facility\&.23\&.stats\&.processed
source\&.facility\&.3\&.stats\&.processed
source\&.facility\&.12\&.stats\&.processed
source\&.facility\&.4\&.stats\&.processed
source\&.facility\&.13\&.stats\&.processed
source\&.facility\&.5\&.stats\&.processed
source\&.facility\&.14\&.stats\&.processed
source\&.facility\&.6\&.stats\&.processed
source\&.facility\&.15\&.stats\&.processed
source\&.facility\&.other\&.stats\&.processed
global\&.payload_reallocs\&.stats\&.processed
global\&.msg_clones\&.stats\&.processed
global\&.sdata_updates\&.stats\&.processed
tag\&.\&.source\&.s_tcp\&.stats\&.processed
.fi
.if n \{\
.RE
.\}
.SH "DISPLAYING METRICS AND STATISTICS"
.HP \w'\fBsyslog\-ng\-query\ sum\fR\ 'u
\fBsyslog\-ng\-query sum\fR [options]
.PP
The
\fBsyslog\-ng\-query sum <query>\fR
command lists the nodes that match the query, and their values\&. For example, the
\fBsyslog\-ng\-query sum "destination*"\fR
command lists the configured destinations, and the metrics related to each destination\&. An example output:
.sp
.if n \{\
.RS 4
.\}
.nf
destination\&.java\&.d_elastic#0\&.java_dst(ElasticSearch,elasticsearch\-syslog\-ng\-test,t7cde889529c034aea9ec_micek)\&.stats\&.dropped: 0
destination\&.java\&.d_elastic#0\&.java_dst(ElasticSearch,elasticsearch\-syslog\-ng\-test,t7cde889529c034aea9ec_micek)\&.stats\&.processed: 0
destination\&.java\&.d_elastic#0\&.java_dst(ElasticSearch,elasticsearch\-syslog\-ng\-test,t7cde889529c034aea9ec_micek)\&.stats\&.stored: 0
destination\&.d_elastic\&.stats\&.processed: 0
.fi
.if n \{\
.RE
.\}
.PP
The
\fBsyslog\-ng\-query sum\fR
command has the following options:
.PP
\fB\-\-aggregate\fR
.RS 4
Add up the result of each matching node and return only a single number\&.
.RE
.PP
For example, the
\fBsyslog\-ng\-query sum \-\-aggregate "destination*\&.dropped"\fR
command displays the number of messages dropped by the instance\&.
.SH "FILES"
.PP
/opt/syslog\-ng/sbin/syslog\-ng\-query
.SH "SEE ALSO"
.PP
\m[blue]\fB\fBThe syslog\-ng Administrator Guide\fR\fR\m[]
.PP
\fBsyslog\-ng\&.conf\fR(5)
.PP
\fBsyslog\-ng\fR(8)
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
For the detailed documentation of see
\m[blue]\fB\fBThe 5 F6 Administrator Guide\fR\fR\m[]
.PP
If you experience any problems or need help with syslog\-ng, visit
\m[blue]\fBvisit the syslog\-ng wiki\fR\m[]
or the
\m[blue]\fB\fBsyslog\-ng mailing list\fR\fR\m[]\&.
.PP
For news and notifications about of syslog\-ng, visit the
\m[blue]\fB\fBsyslog\-ng Blog\fR\fR\m[]\&.
.sp .5v
.RE
.SH "AUTHOR"
.PP
This manual page was written by the BalaBit Documentation Team <documentation@balabit\&.com>\&.
.SH "COPYRIGHT"


syslog-ng-query/syslog-ng-query.c
/*
 * Copyright (c) 2002-2015 BalaBit
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include "syslog-ng.h"
#include "gsocket.h"
#include "control_client.h"
#include "cfg.h"
#include "reloc.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#if HAVE_GETOPT_H
#include <getopt.h>
#endif

static gchar *control_name = PATH_CONTROL_SOCKET;
static ControlClient *control_client;

void usage(const gchar *bin_name);

static gboolean
slng_send_cmd(const gchar * const cmd)
{
  if (!control_client_connect(control_client))
    {
      return FALSE;
    }

  if (control_client_send_command(control_client,cmd) < 0)
    {
      return FALSE;
    }

  return TRUE;
}

static GString *
slng_run_command(const gchar * const command)
{
  if (slng_send_cmd(command))
    return control_client_read_reply(control_client);

  return NULL;
}

static gboolean sum_aggr = FALSE;

static GOptionEntry sum_options[] =
{
  { "aggregate", 'a', 0, G_OPTION_ARG_NONE, &sum_aggr,
      "aggregate all values to a single number", NULL },
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};

static GString *_sum_command_builder(const gchar *path) {
  GString *cmd;
  const gchar *subcmd;

  if (sum_aggr)
    subcmd = "SUM_AGGREGATE";
  else
    subcmd = "SUM";

  cmd = g_string_new("");
  g_string_printf(cmd, "QUERY %s %s\n", subcmd, path);
  return cmd;
}

static gint
slng_sum (int argc, char *argv[], const gchar *mode)
{
  GString *rsp, *cmd;

  if (argc != 2)
    {
      fprintf (stderr, "error: need a path argument\n");
      usage (argv[0]);
      return 1;
    }

  cmd = _sum_command_builder(argv[1]);
  rsp = slng_run_command (cmd->str);
  g_string_free(cmd, TRUE);
  if (rsp == NULL)
    return 1;

  printf("%s", rsp->str);
  if (rsp->len > 0)
    printf("\n");

  g_string_free(rsp, TRUE);

  return 0;
}

static gint
_slng_list (const gchar * path)
{
  GString *cmd = g_string_new("");
  g_string_printf(cmd, "QUERY LIST %s\n", path);
  GString *rsp = slng_run_command (cmd->str);
  g_string_free(cmd, TRUE);

  if (rsp == NULL)
    return 1;

  printf ("%s", rsp->str);
  if (rsp->len > 0)
    printf("\n");

  g_string_free (rsp, TRUE);

  return 0;
}

static gint
slng_list (int argc, char *argv[], const gchar *mode)
{
  if (argc == 1) {
      return _slng_list("*");
  } else if (argc == 2) {
      return _slng_list(argv[1]);
  } else {
      fprintf (stderr, "error: need a path argument or default to *\n");
      usage (argv[0]);
      return 1;
    }
}

static GOptionEntry no_options[] =
{
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};

const gchar *
get_mode(int *argc, char **argv[])
{
  gint i;
  const gchar *mode;

  for (i = 1; i < (*argc); i++)
    {
      if ((*argv)[i][0] != '-')
        {
          mode = (*argv)[i];
          memmove(&(*argv)[i], &(*argv)[i+1], ((*argc) - i) * sizeof(gchar *));
          (*argc)--;
          return mode;
        }
    }
  return NULL;
}

static GOptionEntry slng_options[] =
{
  { "control", 'c', 0, G_OPTION_ARG_STRING, &control_name,
    "syslog-ng control socket", "<socket>" },
  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
};

static struct
{
  const gchar *mode;
  const GOptionEntry *options;
  const gchar *description;
  gint (*main)(gint argc, gchar *argv[], const gchar *mode);
} modes[] =
  {
    { "list", no_options, "lists the available property names with "
        "full path/FQDN, but without the top root node", slng_list },
    { "sum", sum_options, "summarize values stored in nodes that are matched "
        "to the filter/query result", slng_sum },
    { NULL, NULL },
  };

static struct
{
  const gchar *query;
  const gchar *description;
} queries[] =
  {
    { "sum \"destination.tcp.d_network#0.10.50.0.111.514.stored\"",
        "number of messages stored in the given message queue" },
    { "sum 'destination.tcp.d_network#?.*.stored'",
        "messages stored in each message queue and a sum" },
    { "sum --aggregate 'destination.tcp.d_network#0.10.50.0.111.514.stored'",
        "only output total sum as a single number" },
    { NULL, NULL }
  };

void
usage(const gchar *bin_name)
{
  gint mode, query;

  fprintf(stderr, "Syntax: %s <command> [options]\nPossible commands are:\n", bin_name);
  for (mode = 0; modes[mode].mode; mode++)
    {
      fprintf(stderr, "    %-20s %s\n", modes[mode].mode, modes[mode].description);
    }

  fprintf(stderr, "Query examples:\n");
  for (query = 0; queries[query].query; query++)
    {
      fprintf(stderr, "    %s\n            %s\n", queries[query].query, queries[query].description);
    }
  exit(1);
}

#include "reloc.c"

int
main(int argc, char *argv[])
{
  const gchar *mode_string;
  GOptionContext *ctx;
  gint mode;
  GError *error = NULL;
  int result;

  control_name = get_reloc_string(PATH_CONTROL_SOCKET);

  mode_string = get_mode(&argc, &argv);
  if (!mode_string)
    {
      usage(argv[0]);
    }

  ctx = NULL;
  for (mode = 0; modes[mode].mode; mode++)
    {
      if (strcmp(modes[mode].mode, mode_string) == 0)
        {
          ctx = g_option_context_new(mode_string);
          #if GLIB_CHECK_VERSION (2, 12, 0)
          g_option_context_set_summary(ctx, modes[mode].description);
          #endif
          g_option_context_add_main_entries(ctx, modes[mode].options, NULL);
          g_option_context_add_main_entries(ctx, slng_options, NULL);
          break;
        }
    }
  if (!ctx)
    {
      fprintf(stderr, "Unknown command\n");
      usage(argv[0]);
    }

  if (!g_option_context_parse(ctx, &argc, &argv, &error))
    {
      fprintf(stderr, "Error parsing command line arguments: %s\n", error ? error->message : "Invalid arguments");
      g_clear_error(&error);
      g_option_context_free(ctx);
      return 1;
    }
  g_option_context_free(ctx);

  control_client = control_client_new(control_name);

  result = modes[mode].main(argc, argv, modes[mode].mode);
  control_client_free(control_client);
  return result;
}
